### PR TITLE
Don't mutate object when subbing links

### PR DESCRIPTION
**Pull request by asantas93**
This is based on the changes in https://github.com/CitrineInformatics/taurus/pull/59. If we merge both, we ought to be able to simplify some serde logic in citrine-python (and almost certainly taurus as well).

### CODE REVIEW

File: taurus/client/json_encoder.py [hunk 1]
```diff
@@ -153,7 +153,7 @@ def thin_dumps(obj, **kwargs):
     """
     set_uuids(obj)
     res = deepcopy(obj)
```
Comment (bfolie) [inline taurus/client/json_encoder.py]: Can we get rid of this deepcopy now that `substitute_links` does not mutate?

File: taurus/util/impl.py [hunk 2]
```diff
@@ -25,19 +26,57 @@ def func(base_obj):
     return
 
 
+def _substitute(thing, sub, applies, visited: Dict[int, object] = None):
```
Comment (bfolie) [inline taurus/util/impl.py]: Could you add a docstring for this method?
Also, why is the typehint for `visited` a Dict from int to object? Shouldn't it be `Dict[object, object]`?

Comment (asantas93) [reply taurus/util/impl.py]: yeah that's from an old implementation, I will add a docstring and fix the type hint

File: taurus/client/json_encoder.py (hunk 1 above)
Comment (asantas93) [reply taurus/client/json_encoder.py]: probably, I'll look around for `deepcopy`s to remove

**Pull request merged**